_posts/2023-06-22-a-simple-rest-api-with-play-framework-and-scala.md

```diff
@@ -0,0 +1,420 @@
+
+We also need to describe the correspondence between URIs and function calls on our controller. This is achieved through the routes file:
+![alt "Routes structure"](../images/play-rest-api/routes_structure.jpeg)
+
+```scala
+GET     /                 controllers.HomeController.index()
+GET     /car/:id          controllers.HomeController.get(id: Long)
+GET     /cars             controllers.HomeController.getAll()
+
++ nocsrf
+DELETE  /car/delete/:id    controllers.HomeController.delete(id: Long)
++ nocsrf
+PUT     /car/update        controllers.HomeController.update()
++ nocsrf
+POST    /car/save          controllers.HomeController.save()
+```
+Here we define the HTTP method for a specific URI and the method from our controllers layer that will get called when the URI is hit.
+
+In general the Action components of Play are a very flexible way to cover from the default functionality to more advanced.
+
+## 3.4 Application Loader
+Finally we need to put all of the above together and define our main application entry.
+![alt "Application Loader"](../images/play-rest-api/application_loader.jpeg)
+
+Here we define and initialize our main entry point. Initialize all needed components like Managers, Repositories, Routers etc. and finally return an instance of our application. 
+
+```scala
+import controllers.HomeController
+import managers.CarManager
+import persistence.InMemoryPersistence
+import play.api.mvc.EssentialFilter
+import play.filters.HttpFiltersComponents
+import play.api.routing.Router
+import play.api.{Application, ApplicationLoader, BuiltInComponentsFromContext}
+import play.filters.csrf.CSRF
+import repositories.CarRepository
+import router.Routes
+
+
+class RestApplicationLoader extends ApplicationLoader{
+  def load(context: ApplicationLoader.Context): Application = {
+    val components = new RestApiComponents(context)
+
+    components.application
+  }
+}
+
+class RestApiComponents(context: ApplicationLoader.Context)
+  extends BuiltInComponentsFromContext(context) with HttpFiltersComponents{
+
+  lazy val carRepository: CarRepository = new CarRepository(InMemoryPersistence)
+  lazy val carManager: CarManager = new CarManager(carRepository)
+  lazy val homeController: HomeController = new HomeController(carManager, controllerComponents)
+  lazy val router: Router = new Routes(httpErrorHandler, homeController, "/")
+
+  override def httpFilters: Seq[EssentialFilter] = super.httpFilters
+}
+```
+
+We could have different versions of our application and switch between them. This is achieved by the application.conf file where we define which is the main ApplicationLoader class.
+
+```conf
+play.application.loader=RestApplicationLoader
+
+play.filters {
+    enabled += "play.filters.csrf.CSRFFilter"
+}
+```
+
+In this simple example we initialize all needed classes "by hand" but as the project is getting more complex we may want to switch to [Dependency Injection](https://en.wikipedia.org/wiki/Dependency_injection) framework.
```

Comment:
Isn't DI what we're already doing in the code?
This sentence implies that there's a "deeper" DI that we can use. What is it, and how we can use it?